test/SILOptimizer/sil_combine_global_addr.sil

```diff
@@ -0,0 +1,106 @@
+// RUN: %target-sil-opt -assume-parsing-unqualified-ownership-sil -enforce-exclusivity=none -enable-sil-verify-all %s -sil-combine -verify-skip-unreachable-must-be-last -devirtualizer | %FileCheck %s
+
+// Test to see if concrete type can be propagated from
+// global_addr in sil_combiner.
+sil_stage canonical
+
+import Builtin
+import Swift
+protocol SomeProtocol {
+  func foo() -> Int
+}
+class SomeClass  : SomeProtocol {
+  init()
+  func foo() -> Int
+}
+sil hidden [thunk] [always_inline] @foo_ : $@convention(witness_method:SomeProtocol) (@in_guaranteed SomeClass) -> Int {
+bb0(%0 : $*SomeClass):
+  %1 = load %0 : $*SomeClass
+  %2 = class_method %1 : $SomeClass, #SomeClass.foo!1 : (SomeClass) -> () -> Int, $@convention(method) (@guaranteed SomeClass) -> Int
+  %3 = apply %2(%1) : $@convention(method) (@guaranteed SomeClass) -> Int
+  return %3 : $Int
+}
+sil hidden [thunk] [always_inline] @foo : $@convention(method) (@guaranteed SomeClass) -> Int {
+bb0(%0: $SomeClass):
+  %1 = integer_literal $Builtin.Int64, 10
+  %2 = struct $Int (%1 : $Builtin.Int64)
+  return %2 : $Int
+}
+sil_global hidden [let] @$global_var : $SomeProtocol
+
+// CHECK-LABEL: sil @witness_global_addr
+// CHECK: bb0
+// CHECK: alloc_global 
+// CHECK: global_addr
+// CHECK: init_existential_addr
+// CHECK: alloc_ref
+// CHECK: store
+// CHECK: function_ref
+// CHECK: apply
+// CHECK: return
```

Comment:
Quick suggestion. Rather than checking for return to signal the end of the function, can you check for:

```
} // end sil function '$INSERT_FUNCTION_NAME_HERE'
```

This prevents patterns from matching in the wrong function in an absolute way. Can you change the patterns here to use that?